### PR TITLE
SAA-109 intial WIP commit for retrieval of activity schedules endpoint.

### PR DESCRIPTION
**Pull request by matt-crowson-moj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Activity.kt [hunk 1]
```diff
@@ -61,4 +62,21 @@ data class Activity(
   val createdTime: LocalDateTime,
 
   val createdBy: String
-)
+) {
+  fun isActiveOn(date: LocalDate) = active && startDate <= date && (endDate == null || date.isBefore(endDate))
+
+  override fun equals(other: Any?): Boolean {
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Activity.kt:68]: Overrides needed to allow for JPA circular dependencies. New integration test highlighted this.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ActivityScheduleService.kt [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      .groupBy { it.activitySchedule }
+      .map { (schedule, instances) -> schedule.copy(instances = instances.toMutableList()) }
+
+    return transform(schedulesWithInstancesOnChosenDateAndSlot)
+  }
+
+  private fun List<ActivitySchedule>.selectSchedulesWithActiveActivitiesOn(date: LocalDate) =
+    filter { it.activity.isActiveOn(date) }
+
+  // TODO need to filter on time slot morning/afternoon/evening
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/ActivityScheduleService.kt:34]: As the comment suggests.  This is not checking the timeslot so all instances morning and afternoon will be returned at the moment.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Activity.kt [hunk 3]
```diff
@@ -61,4 +62,21 @@ data class Activity(
...
+    if (this === other) return true
+    if (other == null || Hibernate.getClass(this) != Hibernate.getClass(other)) return false
+    other as Activity
+
+    return activityId != null && activityId == other.activityId
+  }
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Activity.kt:74]: Might the equals check on the ID column give different results before and after persisting? I would have thought the equals method should exclude ID columns for that purpose, but might be wrong!

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/helpers/ExtensionFunctions.kt [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.helpers
+
+import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.model.ActivitySchedule
+import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.model.Allocation
+
```
Comment (timharrison-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/helpers/ExtensionFunctions.kt:5]: I generally think the best place for extension functions, particularly though related to our own classes, might be near to the classes, or in them? Just a thought and no biggy.

**Review by timharrison-moj — APPROVED** [review]
Couple of questions, but LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Activity.kt (hunk 3 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/Activity.kt:74]: So this is actually auto-generated by IntelliJ and the JPA buddy plugin.  

Based on [this]( https://github.com/jpa-buddy/kotlin-entities/blob/main/src/main/kotlin/com/jpabuddy/kotlinentities/domain.kt#L118) on Github [here](https://github.com/jpa-buddy/kotlin-entities)

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/helpers/ExtensionFunctions.kt (hunk 4 above)
Comment (matt-crowson-moj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/helpers/ExtensionFunctions.kt:5]: This should probably at least have been in the integration test folder thinking about it.  Will move in next PR.

**Pull request merged**